### ISSUE TITLE
system-probe container cannot be killed

### ISSUE THREAD

**Issue by ploxiln** [DataDog/helm-charts#1084]
**Describe what happened:**

We use this helm chart to deploy the datadog-agent to our kubernetes clusters. When rolling out changes to the values.yaml, which updates some configuration of the daemonset and triggers a rolling update of the agent pods on each node, the old pods are stuck in Terminating state for 15+ minutes, dramatically slowing down the roll-out, and causing a 15+ minute stats/logs outage on a k8s node:

```
$ rodeo aws exec staging -- kubectl get pods -l app.kubernetes.io/component=agent
[rodeo][info] assuming AWS role oktaReadOnlyRole in account 385139207480 ...
NAME            READY   STATUS        RESTARTS   AGE
datadog-4sr52   4/4     Running       0          25d
datadog-5dtv6   4/4     Running       0          25d
datadog-8bbbt   4/4     Running       0          25d
datadog-8vfzb   4/4     Running       0          25d
datadog-cv76f   4/4     Running       0          9d
datadog-gxm9q   4/4     Running       0          25d
datadog-jstv4   4/4     Running       0          25d
datadog-n5zzg   4/4     Running       0          25d
datadog-p9sfj   4/4     Running       0          7d10h
datadog-pb8v8   0/4     Terminating   0          25d
datadog-rchrg   4/4     Running       0          25d
datadog-scw9k   4/4     Running       0          25d
datadog-sj7jb   4/4     Running       0          5d6h
datadog-sxxlt   4/4     Running       0          25d
datadog-wg6k2   0/4     Terminating   2          9d
```

This is because of the `system-probe` container:

> Warning FailedKillPod 5m17s kubelet error killing pod: failed to "KillContainer" for "system-probe" with KillContainerError: "rpc error: code = Unknown desc = Error response from daemon: cannot stop container: c59e2bbf25ae89776a9022f5b5270a89d0ff4208c35d3041b2a7e3c12c9ff345: container c59e2bbf25ae PID 25317 is zombie and can not be killed. Use the --init option when creating containers to run an init inside the container that forwards signals and reaps processes"

**Describe what you expected:**

The pod should terminate promptly so it can be replaced with updated configuration.

**Steps to reproduce the issue:**

Have a k8s cluster with this chart deployed, configured to enable system-probe. Update some values, deploy updated chart to cause daemonset to be updated, to cause rolling replacement of pods.

**Additional environment details (Operating System, Cloud provider, etc):**

kubernetes-version: v1.23.17-eks-0a21954 (yup, using AWS EKS)
datadog-chart-version: 3.24.0
datadog-agent-version: 7.43.1 (the default for the chart)
datadog stuff enabled: a lot ...

**Comment by ploxiln** [comment]
We think we were able to fix this with a work-around: set `agents.shareProcessNamespace: true`

With this enabled, the kubernetes "pause" container becomes the PID 1 init process for all the datadog-agent pod containers, and it reaps child processes including the system-probe (which somehow remained for 10+ minutes as a zombie process in our case). This seems to enable the datadog-agent to be terminated cleanly and promptly.

This does require us to set `datadog.useHostPID: false`, but we weren't using that anyway (we were just using udp/tcp, not using unix-domain sockets, which are required for the useHostPID thing to work).

**Comment by ploxiln** [comment]
Eh I take it back, that didn't fix it, still having datadog-agent pods take 25+ minutes to terminate because the system-probe container is stuck and can't be killed.

**Comment by ploxiln** [comment]
I have a new theory: this system-probe is creating special objects in the kernel, and when it's time for the kernel to clean them up as the process terminates, that's what takes forever, because it's an order of magnitude more than the kernel expected for this kind of object, and they're managed with RCU in the kernel (a locking/sychronization mechanism that optimizes for multi-core read performance at the expense of update performance). Maybe this object is ... bpf programs? kprobes? something that is created per-container or per-process? (we have loads of containers on our k8s nodes ...)

**Comment by ploxiln** [comment]
(I _think_ we've now solved this problem, by disabling NPM, while keeping USM enabled ...)

**Comment by ploxiln** [comment]
(I take that back, the issue persists. It just apparently takes a few days to build up the state that makes it take 25+ minutes to terminate. I think that disabling USM in order to eliminate the system-probe container would avoid it, but my coworkers really value the USM and want to keep it.)

**Comment by ploxiln** [comment]
I haven't been able to find a clear cause yet, not quite familiar enough to quickly inspect mounts and such, but here's proc status of a system-probe process that's been a zombie for about 20 minutes:

```
$ cat /proc/19099/status
Name:   system-probe
State:  Z (zombie)
Tgid:   19099
Ngid:   0
Pid:    19099
PPid:   18424
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
FDSize: 0
Groups: 0
NStgid: 19099   107
NSpid:  19099   107
NSpgid: 19099   107
NSsid:  19099   107
Threads:        2
SigQ:   0/30446
SigPnd: 0000000000000000
ShdPnd: 0000000000000100
SigBlk: fffffffc3bba2a00
SigIgn: 0000000000000000
SigCgt: fffffffd7fc1feff
CapInh: 0000000000000000
CapPrm: 00000000a92c7dff
CapEff: 00000000a92c7dff
CapBnd: 00000000a92c7dff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Seccomp_filters:        0
Speculation_Store_Bypass:       thread vulnerable
Cpus_allowed:   ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        578
nonvoluntary_ctxt_switches:     2
```

(... yeah, I'm thinking this is nothing to do with the chart, but a behavior of the agent. I should upgrade to latest and open a report in the agent repo I guess.)

**Comment by ploxiln** [comment]
But while I'm here ... this thing has two threads, and here's the other thread, stuck in disk sleep! must be a weird mount in the container

```
[ssm-user@ip-10-2-56-149 ~]$ cat /proc/19099/task/19500/status
Name:   system-probe
State:  D (disk sleep)
Tgid:   19099
Ngid:   0
Pid:    19500
PPid:   18424
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
FDSize: 0
Groups: 0
NStgid: 19099   107
NSpid:  19500   137
NSpgid: 19099   107
NSsid:  19099   107
Threads:        2
SigQ:   0/30446
SigPnd: 0000000000000000
ShdPnd: 0000000000000100
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: fffffffd7fc1feff
CapInh: 0000000000000000
CapPrm: 00000000a92c7dff
CapEff: 00000000a92c7dff
CapBnd: 00000000a92c7dff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Seccomp_filters:        0
Speculation_Store_Bypass:       thread vulnerable
Cpus_allowed:   ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        7476740
nonvoluntary_ctxt_switches:     7793
```

**Comment by guyarb** [comment]
Hey @ploxiln, sorry for the delay.

I do see you're using `7.43.1` version, and unfortunately one of the core components of USM had an issue (affected versions `7.42.0`-7.44.0`, solved in `7.44.1`) that could lead for such a scenario. 
I do suggest to upgrade the agent to `7.44.1` (and if possible to the latest version of `7.47.0`).
If that doesn't work, could you open a support case with us? That will allow you to send us a flare which will give us fine grained information about the environment you're running in which should let us reproduce a bit more easily.

**Comment by ploxiln** [comment]
Makes sense, thanks, will do that update promptly ...
